src/backend/opencl/blas_clblast.cpp

```diff
@@ -0,0 +1,153 @@
+
+#if defined(USE_CLBLAST)
+
+#include <complex>
+
+#include <blas.hpp>
+#include <Array.hpp>
+#include <err_opencl.hpp>
+#include <math.hpp>
+#include <transpose.hpp>
+#include <arith.hpp>
+#include <reduce.hpp>
+#include <complex.hpp>
+
+#include <clblast.h>
+#include <err_clblast.hpp>
+
+#if defined(WITH_OPENCL_LINEAR_ALGEBRA)
+#include <cpu/cpu_blas.hpp>
+#endif
+
+namespace opencl
+{
+
+void
+initBlas()
+{
+  // Nothing to do here for CLBlast
+}
+
+clblast::Transpose
+toClblastTranspose(af_mat_prop opt)
+{
+    switch(opt) {
+        case AF_MAT_NONE    : return clblast::Transpose::kNo;
+        case AF_MAT_TRANS   : return clblast::Transpose::kYes;
+        case AF_MAT_CTRANS  : return clblast::Transpose::kConjugate;
+        default             : AF_ERROR("INVALID af_mat_prop", AF_ERR_ARG);
+    }
+}
+
+// Defines type conversions from ArrayFire (OpenCL) to CLBlast (C++ std)
+template <typename T> struct CLBlastType { using Type = T; };
+template <> struct CLBlastType<cfloat> { using Type = std::complex<float>; };
+template <> struct CLBlastType<cdouble> { using Type = std::complex<double>; };
+
+// Converts a constant from ArrayFire types (OpenCL) to CLBlast types (C++ std)
+template <typename T> typename CLBlastType<T>::Type toCLBlastConstant(const T val);
+
+// Specializations of the above function
+template <> float toCLBlastConstant(const float val) { return val; }
+template <> double toCLBlastConstant(const double val) { return val; }
+template <> std::complex<float> toCLBlastConstant(cfloat val) { return {val.s[0], val.s[1]}; }
+template <> std::complex<double> toCLBlastConstant(cdouble val) { return {val.s[0], val.s[1]}; }
+
+template<typename T>
+Array<T> matmul(const Array<T> &lhs, const Array<T> &rhs,
+                af_mat_prop optLhs, af_mat_prop optRhs)
+{
+#if defined(WITH_OPENCL_LINEAR_ALGEBRA)
+    if(OpenCLCPUOffload(false)) {   // Do not force offload gemm on OSX Intel devices
+        return cpu::matmul(lhs, rhs, optLhs, optRhs);
+    }
+#endif
+
+    const auto lOpts = toClblastTranspose(optLhs);
+    const auto rOpts = toClblastTranspose(optRhs);
+
+    const auto aRowDim = (lOpts == clblast::Transpose::kNo) ? 0 : 1;
+    const auto aColDim = (lOpts == clblast::Transpose::kNo) ? 1 : 0;
+    const auto bColDim = (rOpts == clblast::Transpose::kNo) ? 1 : 0;
+
+    const dim4 lDims = lhs.dims();
+    const dim4 rDims = rhs.dims();
+    const int M = lDims[aRowDim];
+    const int N = rDims[bColDim];
+    const int K = lDims[aColDim];
+
+    Array<T> out = createEmptyArray<T>(af::dim4(M, N, 1, 1));
+    const auto alpha = scalar<T>(1);
+    const auto beta  = scalar<T>(0);
+    const auto alpha_clblast = toCLBlastConstant(alpha);
+    const auto beta_clblast = toCLBlastConstant(beta);
+
+    const dim4 lStrides = lhs.strides();
+    const dim4 rStrides = rhs.strides();
+    if(rDims[bColDim] == 1) {
+        CLBLAST_CHECK(
+            clblast::Gemv(clblast::Layout::kColMajor, lOpts,
```

Comment:
So you mean that the `matmul` function can just call either `gpu_blas_gemv_func` or `gpu_blas_gemm_func`? If that's the case then we can merge `blas_clblast.cpp` and `blas_clblas.cpp` into one function. I guess that's what you mean with minimising the specialised logic? Yes, I think that's a good idea!